integration-tests/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+///
+/// If just outputs non utf-8 bytes or can't be called or fails this
+/// will panic.
+///
+/// This will capture stdout, but not stderr so warnings, errors, traces
+/// and similar will be printed like normal. In case it fails it will also
+/// print the previously captured stdout.
+pub fn just(args: &[&str]) -> Result<String, anyhow::Error> {
+    let Output { status, stdout, .. } = Command::new("just")
+        .args(args)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::inherit())
+        .output()?;
+
+    let output = String::from_utf8(stdout)?;
+    if status.success() {
+        Ok(output)
+    } else {
+        bail!("Cmd `just` failed! Output:\n{output}Just Exit Status: {status}");
+    }
+}
+
+pub async fn send_assert_success(client: &Client, req: Request) {
+    let method = req.method().clone();
+    let target = req.url().clone();
+    let response = client.execute(req).await.unwrap();
+    let status = response.status();
+    if !status.is_success() {
+        let bytes = response.bytes().await.unwrap();
+        let text = String::from_utf8_lossy(&bytes);
+        panic!("Failed to {method} {target}, status {status}, body: {text}");
+    }
+}
+
+pub async fn send_assert_success_json<O>(client: &Client, req: Request) -> O
+where
+    O: DeserializeOwned,
+{
+    let method = req.method().clone();
+    let target = req.url().clone();
+    let response = client.execute(req).await.unwrap();
+    let status = response.status();
+    if !status.is_success() {
+        let bytes = response.bytes().await.unwrap();
+        let text = String::from_utf8_lossy(&bytes);
+        panic!("Failed to {method} {target}, status {status}, body: {text}");
+    }
+
+    let bytes = response.bytes().await.unwrap();
+
+    match serde_json::from_slice::<O>(&bytes) {
+        Ok(out) => out,
+        Err(err) => {
+            let text = String::from_utf8_lossy(&bytes);
+            panic!("Failed to decode body of {method} {target}, error: {err}\nbody: {text}")
+        }
+    }
+}
+
+pub async fn test_app<A, F>(
+    configure: impl FnOnce(&mut Table),
+    test: impl FnOnce(Arc<Client>, Arc<Url>, Services) -> F,
+) where
+    F: Future<Output = Result<(), Panic>>,
+    A: Application + 'static,
+{
+    clear_env();
+    start_test_service_containers().unwrap();
+
+    let services = create_web_dev_services().await.unwrap();
+
+    let mut config = toml! {
+        [storage.postgres]
+        [storage.elastic]
+        [embedding]
+        directory = "../assets/smbert_v0003"
+    };
+
+    let storage = config["storage"].as_table_mut().unwrap();
+    storage["postgres"]
+        .as_table_mut()
+        .unwrap()
+        .insert("base_url".into(), services.postgres.clone().into());
+    let elastic = storage["elastic"].as_table_mut().unwrap();
+    let (url, index) = services.elastic_search.rsplit_once('/').unwrap();
+    elastic.insert("url".into(), url.into());
+    elastic.insert("index".into(), index.into());
```

Comment:
I'm not the biggest fan how this turned out wrt. readability, but for now that is good enough I guess.